pkg/client/typed/discovery/discovery_client.go

```diff
@@ -209,75 +211,93 @@ func IsGroupDiscoveryFailedError(err error) bool {
 
 // serverPreferredResources returns the supported resources with the version preferred by the
 // server. If namespaced is true, only namespaced resources will be returned.
-func (d *DiscoveryClient) serverPreferredResources(namespaced bool) ([]schema.GroupVersionResource, error) {
+func (d *DiscoveryClient) serverPreferredResources(namespaced bool) ([]*unversioned.APIResourceList, error) {
 	// retry in case the groups supported by the server change after ServerGroup() returns.
 	const maxRetries = 2
 	var failedGroups map[schema.GroupVersion]error
-	var results []schema.GroupVersionResource
-	var resources map[schema.GroupResource]string
+	var grVersions map[schema.GroupResource]string                              // selected version of a GroupResource
+	var grApiResources map[schema.GroupResource]*unversioned.APIResource        // selected APIResource for a GroupResource
+	var gvApiResourceLists map[schema.GroupVersion]*unversioned.APIResourceList // blueprint for a APIResourceList for later grouping
+	var result []*unversioned.APIResourceList
 RetrieveGroups:
 	for i := 0; i < maxRetries; i++ {
-		results = []schema.GroupVersionResource{}
-		resources = map[schema.GroupResource]string{}
+		grVersions = map[schema.GroupResource]string{}
+		grApiResources = map[schema.GroupResource]*unversioned.APIResource{}
+		gvApiResourceLists = map[schema.GroupVersion]*unversioned.APIResourceList{}
+		result = []*unversioned.APIResourceList{}
 		failedGroups = make(map[schema.GroupVersion]error)
 		serverGroupList, err := d.ServerGroups()
 		if err != nil {
-			return results, err
+			return result, err
 		}
 
 		for _, apiGroup := range serverGroupList.Groups {
 			versions := apiGroup.Versions
 			for _, version := range versions {
```

Comment:
if I understand the goal correctly, you want the best version for every GR.  I would traverse the versions in reverse order, and stomp them in a map.  Last one wins, then build the return in a random order per group since no group will have duplicates.